lib/ecto/type.ex

```diff
@@ -928,10 +928,8 @@ defmodule Ecto.Type do
     end
   end
 
-  defp validate_decimal({:ok, %Decimal{coef: coef}}, value) when coef in [:inf, :qNaN, :sNaN] do
-    message = "cannot cast #{inspect value} to :decimal. Define custom type to handle NaN or infinite decimals"
-    raise Ecto.CastError, type: :decimal, value: value, message: message
-  end
-
-  defp validate_decimal(other, _), do: other
+  defp validate_decimal({:ok, %Decimal{coef: coef}}) when coef in [:inf, :qNaN, :sNaN],
+    do: :error
```

Comment:
I probably should have asked that on the previous PR, but is the internal structure of the `Decimal` struct and the values we match on in here documented?

Reply:
Yes, they are public API.